---
title: "Generate gff and fasta files for yeast genome plus constructs"
author: "Sam Haynes"
date: "02/07/2021"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(here)

GTF_COLNAMES = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")
FEATURES = c("primary_transcript","Promoter", "CDS", "Terminator", "URA3", "M1", "M2", "M3")
```

```{r generate-plasmid-feature-annotations}
plasmid_features <- read_delim(here("./data/input/Scer_ref_genome/construct_integrated_genome/plasmid_features.csv"), ";")

plasmid_features_long <- plasmid_features %>% 
  select(-Linear_sequence,-Orig_Promoter,-Orig_Terminator) %>% 
  pivot_longer(!Construct_name,names_to = "feature", values_to = "position") %>% 
  separate(feature,into=c("feature","location"),sep="_") %>% 
  pivot_wider(names_from = location, values_from = position) %>%
  filter(!is.na(start))

plasmid_primary_transcript_feature <- plasmid_features_long %>%
  filter(feature != "URA3") %>%
  group_by(Construct_name) %>%
  summarise(feature = "primary_transcript", start = min(start), stop = max(stop))

plasmid_features_all <- plasmid_features_long %>%
  bind_rows(plasmid_primary_transcript_feature) %>%
  mutate(strand = ifelse(feature == "URA3", "-", "+")) %>%
  mutate(source = "chimera-project", score = ".", frame = ".", attribute = paste0("ID=", Construct_name), seqname = Construct_name, end = stop) %>%
  select(GTF_COLNAMES)

genome_features <- read_tsv(here("data/input/Scer_ref_genome/longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff"), comment="#", col_names = GTF_COLNAMES)

output_gff_file <- here("./data/input/Scer_ref_genome/construct_integrated_genome/longest_full_ORF_with_constructs.gff")

write_lines(
    x=c("##gff-version 3", 
        "##three prime UTRs from most_abundant_full-ORF_transcripts_ypd.gff3",
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=output_gff_file)

write_tsv(bind_rows(genome_features,plasmid_features_all), output_gff_file, append = TRUE)

```

```{r create-plasmid-fasta-files}
construct_fasta_file_names <- list.files(here("./data/input/Scer_ref_genome/construct_integrated_genome/linearized_plasmids/"), full.names = TRUE)

construct_fasta <- lapply(construct_fasta_file_names,read_table)

genomic_fasta <- read_table(here("./data/input/Scer_ref_genome/saccharomyces_cerevisiae_R64-2-1_20150113_chrnames.fsa"))

write_tsv(genomic_fasta, path = here("./data/input/Scer_ref_genome/construct_integrated_genome/saccharomyces_cerevisiae_R64_with_construct.fa"), col_names = TRUE)

lapply(construct_fasta, write_tsv, append=TRUE, here("./data/input/Scer_ref_genome/construct_integrated_genome/saccharomyces_cerevisiae_R64_with_construct.fa"), col_names = TRUE)


```